### PR TITLE
WIP: NEXT-20949 - Metrics in Shopware

### PR DESCRIPTION
**Pull request by fschmtt**

### CODE REVIEW

**Review by SebastianFranze — CHANGES_REQUESTED** [review]

File: guides/plugins/plugins/framework/metrics/README.md [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+        if (!$source instanceof SalesChannelApiSource) {
+            return [];
+        }
+        
+        $criteria = new Criteria();
+        $criteria->addAssociation('salesChannel');
+        $criteria->addFilter(new EqualsFilter('salesChannel.id', $source->getSalesChannelId()));
+        
+        $theme = $this->themeRepository->search($criteria, $context)->first();
```
[the remark this answers is not in the document]
Comment (fschmtt) [reply guides/plugins/plugins/framework/metrics/README.md]: Switched to plain SQL.

File: guides/plugins/plugins/framework/metrics/README.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+A single metric in Shopware is represented by the `MetricStruct` class, an immutable object that holds all the data associated to the metric.
+
+Let's have a closer look at its properties and what they're about:
+
+| Property  | Mutator            | Description                                                                                        |
+|-----------|--------------------|----------------------------------------------------------------------------------------------------|
+| `$name`     | `n/a`              | The metric's name.                                                                                 |
+| `$type`     | `::withType()`     | Allowed types are `count`, `gauge` and `histogram` (defined as class constants for easier access). |
+| `$value`    | `::withValue()`    | The value of the metric must be one of <code>bool                                                  |int|float|string</code> and mainly depends on the metric's type.                |
+| `$metadata` | `::withMetadata()` | The metadata are additional information that usually are the same for all metrics.                 |
```
[the remark this answers is not in the document]
Comment (fschmtt) [reply guides/plugins/plugins/framework/metrics/README.md:51]: i also like it better to have those both "collections" separated from each other, so the clients can easily distinguish between them. and they can ultimately decide how to handle them - there might be clients that simply merge both, others might want to transform the metatag's keys and prefix them or sth.

there's quite a variety of possible clients/backends that i would like to give implementations the possibility to handle those data differently.

from a technical point of view, yes, both seem to be the same, but as metrics are more likely to be "business" metrics, we should also keep this in mind, when designing the interfaces IMO.

File: guides/plugins/plugins/framework/metrics/README.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+And this is where metrics come into play!
+
+To understand this article you should be familiar with the plugin fundamentals:
+
+{% page-ref page="../../plugin-base-guide.md" %}
+
+## Collecting Metrics in Shopware
+There are several ways of collecting metrics by implementing different interfaces provided by Shopware.
+For event based metrics, e.g. when an order is placed, a simple event subscriber can be used.
+For more resource intensive metrics it's better to use a collector, that is run once a day by scheduled task.
```
Comment (xPand4B) [inline guides/plugins/plugins/framework/metrics/README.md:22]:
```suggestion
For more resource intensive metrics it's better to use a collector, that will run once a day by scheduled task.
```

File: guides/plugins/plugins/framework/metrics/README.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+### Metric struct
+A single metric in Shopware is represented by the `MetricStruct` class, an immutable object that holds all the data associated to the metric.
+
+Let's have a closer look at its properties and what they're about:
+
+| Property  | Mutator            | Description                                                                                        |
+|-----------|--------------------|----------------------------------------------------------------------------------------------------|
+| `$name`     | `n/a`              | The metric's name.                                                                                 |
+| `$type`     | `::withType()`     | Allowed types are `count`, `gauge` and `histogram` (defined as class constants for easier access). |
+| `$value`    | `::withValue()`    | The value of the metric must be one of <code>bool                                                  |int|float|string</code> and mainly depends on the metric's type.                |
```
Comment (xPand4B) [inline guides/plugins/plugins/framework/metrics/README.md:50]: This one has 4 columns in a 3 column table 🤔

File: guides/plugins/plugins/framework/metrics/README.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+| `$metadata` | `::withMetadata()` | The metadata are additional information that usually are the same for all metrics.                 |
+| `$tags`     | `::withTags()`     | Tags are specific to a metric and are typically used for filtering and aggregating metrics.        |
+
+So, when to use tags or metadata? When fetching data and creating a new metric that you want to have additional key-value data specific to this very metric, you'll want to use tags.
+Metadata, on the other hand, are data that are the same for every metric and therefore also added to all metrics only once before dispatching.
+To provide additional custom metadata, see the corresponding section later in this article.
```
Comment (xPand4B) [inline guides/plugins/plugins/framework/metrics/README.md:56]: Maybe add a link to the part mentioned?

File: guides/plugins/plugins/framework/metrics/README.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+use Shopware\Core\System\Metrics\MetricCollection;
+use Shopware\Core\System\Metrics\MetricStruct;
+
+class CustomerRegisteredMetricsSubscriber extends AbstractMetricEventSubscriber
+{
+    private const METRIC_NAME_CUSTOMER_REGISTERED = 'customer.registered';
+
+    public function getSubscribedEvent(): string
+    {
+        CustomerRegisterEvent::class;
```
Comment (xPand4B) [inline guides/plugins/plugins/framework/metrics/README.md:80]: I think the `return` statement is missing

```suggestion
        return CustomerRegisterEvent::class;
```

**Review by xPand4B — CHANGES_REQUESTED** [review]

**Comment by sushmangupta** [comment]
Followup- the feature is yet to be implemented.

**Pull request closed without merging**